sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala

```diff
@@ -116,6 +119,39 @@ trait CodegenSupport extends SparkPlan {
   }
 
   /**
+    * Returns source code to evaluate all the variables, and clear the code of them, to prevent
```

Comment:
Some variables could be generated in the middle of the plan, for example, aggregate, and join, so we can't always use the references of current plan to determine which expression is used or not. So I have two different functions here, we could pass in the used references to the function below.